Support <amp-social-share type="system">

ithinkihaveacat opened the pull request:
Adds support for the [`navigator.share` origin trial](https://developers.google.com/web/updates/2016/10/navigator-share).

This fixes #5799.

**Screenshots**

`<amp-social-share type="system">`:

![adb-screenshot-20170130153712](https://cloud.githubusercontent.com/assets/51244/22429306/16472182-e702-11e6-9e4b-d489839fdfa8.png)

`<amp-social-share type="system" data-mode="replace">`:

![adb-screenshot-20170130153630](https://cloud.githubusercontent.com/assets/51244/22429330/22fb86ac-e702-11e6-81e7-b56677576a6b.png)

**TODO**

* Design/UI questions:
  * [x] Is `system` the right type? (Another possibility: `native`.)
  * [x] Is the MD "share" icon correct? (Note that at the moment this feature is not supported by Safari, and will not appear on iOS.)
  * [x] What to use as the background color?
* [x] "optionally hide all the other buttons if share is supported"
* [x] [Apply for origin trail token](https://docs.google.com/forms/d/e/1FAIpQLSfO0_ptFl8r8G0UFhT0xhV17eabG-erUWBDiKSRDTqEZ_9ULQ/viewform)

**Testing**

To test you need to be running Chrome on a real mobile device. The next step depends on the domain your AMP is hosted on:

* **localhost**
  1. Enable "Experimental Web Platform features" via <chrome://flags/#enable-experimental-web-platform-features>.
* **Other domains**
  1. [Apply for an origin trial token for your domain](https://goo.gl/GR5YLI).
  1. Insert the tokens on your site. (Note that until #7122 is deployed, `<meta http-equiv="origin-trial">` is invalid AMP, and will not be served by the cache.)

Origin trial token for `cdn.ampproject.org`:

```
AvyXp7EY65ehpmAu/9JOWZVDuLAUiZWk9Tpp9h+lkQuZzednBdNKkbZkOobQkKXKuGonBGLIOP4xqgNsxQY0xwEAAABZeyJvcmlnaW4iOiAiaHR0cHM6Ly9jZG4uYW1wcHJvamVjdC5vcmc6NDQzIiwgImZlYXR1cmUiOiAiV2ViU2hhcmUiLCAiZXhwaXJ5IjogMTQ4ODQ2ODUxMn0=
```

cramforce commented on extensions/amp-social-share/0.1/amp-social-share.js:
```diff
@@ -55,6 +55,12 @@ class AmpSocialShare extends AMP.BaseElement {
   buildCallback() {
     const typeAttr = user().assert(this.element.getAttribute('type'),
         'The type attribute is required. %s', this.element);
+    const isTypeSupported = typeAttr !== 'system' ||
```
I'd write this as
`typeAttr == 'system' ? !!navigator.share ? true`;

for slightly better readability.

cramforce commented on extensions/amp-social-share/0.1/amp-social-share.js:
```diff
@@ -55,6 +55,12 @@ class AmpSocialShare extends AMP.BaseElement {
         'The type attribute is required. %s', this.element);
+    const isTypeSupported = typeAttr !== 'system' ||
+        navigator.share !== undefined;
+    if (!isTypeSupported) {
```
Actually, lets just make this special
`if (typeAttr == 'system' && !navigator.share)`

No need to overgeneralize if there is only one case. And then please add a comment.

ithinkihaveacat replied:
Done

cramforce commented on extensions/amp-social-share/0.1/amp-social-share.js:
```diff
@@ -78,17 +84,24 @@ class AmpSocialShare extends AMP.BaseElement {
+      dev().assert(navigator.share !== undefined,
+          'navigator.share disappeared.');
+      navigator.share(parseQueryString(href.substr(href.indexOf('?'))));
+      e.preventDefault();
```
Is there a default action? Wondering because it is not called in the else case.

ithinkihaveacat replied:
Can't figure out what problem this was supposed to fix, so removed it.

jridgewell commented:
Ping. Can you rebase off of the current master?

jridgewell commented:
I think we just need a test for the expected `navigator.share` call param.

ithinkihaveacat commented on extensions/amp-social-share/0.1/amp-social-share.js:
```diff
@@ -85,10 +90,16 @@ class AmpSocialShare extends AMP.BaseElement {
+    if (this.shareEndpoint_ === 'share:') {
+      dev().assert(navigator.share !== undefined,
+          'navigator.share disappeared.');
+      navigator['share'](parseQueryString(href.substr(href.indexOf('?'))));
```
The `navigator['share']` (instead of `navigator.share`) formulation is needed to evade a test in gulp's `check-types` task, though presumably there's a better way… (This task runs on Travis; I can't see how to run it locally.)

ithinkihaveacat commented:
@jridgewell Apologies for the delay; I've rebased and addressed @cramforce's comments. There's still quite a few technical and non-technical issues (the checklist in the PR description) that I'm a bit stuck on, and would appreciate some pointers on.

jridgewell commented:
@abeck000 might be able to better decide the UX questions. I'd prefer it if we used the native share icon in iOS, if there's an public SVG we could use.

ithinkihaveacat commented:
@jridgewell `navigator.share` is not supported on iOS right now (iOS users won't see any share icon) and I don't believe there's any Safari equivalent.

jridgewell commented:
Ohh, then yah, that icon's the normal Android one, right? 👍

cramforce commented on extensions/amp-social-share/0.1/amp-social-share-config.js:
```diff
@@ -80,4 +80,11 @@ const BUILTINS = {
     },
   },
+  system: {
+    shareEndpoint: 'share:',
```
Could we call this `navigator-share`? I think that would avoid potential confusion with a reasonable future URL scheme.

ithinkihaveacat replied:
Renamed.

cramforce commented on extensions/amp-social-share/0.1/amp-social-share.css:
```diff
@@ -70,3 +70,7 @@ amp-social-share {
 }
+
+.amp-social-share-system {
+  background-color: deeppink;
```
I'd prefer something very neutral, like black?

ithinkihaveacat replied:
Changed to white on black. (Was originally deeppink to make it clear it hadn't been "designed"!)

cramforce commented on extensions/amp-social-share/0.1/amp-social-share.css:
```diff
@@ -70,3 +70,7 @@ amp-social-share {
   background-color: #25d366;
 }
+
+.amp-social-share-system {
```
I feel like I'm missing something obvious: Where does the share icon come from that is in your example screen?

ithinkihaveacat replied:
It's in [third_party/optimized-svg-icons/amp-social-share-svgs.css](https://github.com/ampproject/amphtml/pull/6688/files#diff-a6dd12bee6d97a0ead117259d76044d4) with the rest of the SVGs—is that what you mean? (Converted from https://material.io/icons/.)

cramforce replied:
Sorry, yes. I keep getting burned by GH's misfeature to suppress some diffs.

abeck000 commented:
i'll take a look at the UI with @ericlindley-g to figure out what the hiding might look like.

abeck000 commented:
UX answers:

- YES MD icon is correct for native Andriod
- Color : No preference on the color, just keep it simple. Either : white icon on black is fine. (OR black icon on white. which ever makes the most sense and is easily customizable)
- YES "optionally hide all the other buttons if share is supported" > Just to clarify - clicking the share icon will reveal the native share sheet with actual actions

cc @jridgewell @cramforce @ericlindley-g

ithinkihaveacat commented:
@abeck000 

> No preference on the color, just keep it simple.

The icon is by default white-on-black (identical to the "email" type). The colors and icon can be customized via CSS.

> YES "optionally hide all the other buttons if share is supported" > Just to clarify - clicking the share icon will reveal the native share sheet with actual actions

Yes, the actual actions will be available on the native share sheet, which is identical to the share sheet displayed by Chrome itself when the user taps Action overflow menu -> "Share...":

![ezgif com-781d40bf96](https://cloud.githubusercontent.com/assets/51244/22428696/e84b300e-e6ff-11e6-8f34-ba6677456247.gif)

(The actions are unlikely to match those displayed on the web page, since the native share sheet only shows actions that can be completed via currently-installed native applications, whereas the web page will show all actions the page author specified.)

The attribute `data-mode=replace` activates the native share component's "hide all other buttons" mode.

jridgewell commented on third_party/optimized-svg-icons/amp-social-share-svgs.css:
```diff
@@ -35,3 +35,7 @@
 }
+
+.amp-social-share-system {
+  background-image: url('data:image/svg+xml;charset=utf8,%3Csvg%20fill=%22#ffffff%22%20height=%2224%22%20viewBox=%220%200%2024%2024%22%20width=%2224%22%20xmlns=%22http://www.w3.org/2000/svg%22%3E%3Cpath%20d=%22M0%200h24v24H0z%22%20fill=%22none%22/%3E%3Cpath%20d=%22M18%2016.08c-.76%200-1.44.3-1.96.77L8.91%2012.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5%201.25.81%202.04.81%201.66%200%20 ...
```
Please update this to `charset=utf-8`, per https://github.com/ampproject/amphtml/pull/6920.

ithinkihaveacat replied:
Done: all instances of `charset=utf8` in this file have been changed to `charset=utf-8`. (Also added a charset to `.amp-social-share-whatsapp`, which was missing it completely.)

jridgewell commented on third_party/optimized-svg-icons/amp-social-share-svgs.css:
```diff
@@ -5,33 +5,37 @@
 
 .amp-social-share-facebook {
-  background-image: url('data:image/svg+xml;charset=utf8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%20512%20512%22%3E%3Cpath%20fill%3D%22%23ffffff%22%20d%3D%22M211.9%20197.4h-36.7v59.9h36.7V433.1h70.5V256.5h49.2l5.2-59.1h-54.4c0%200%200-22.1%200-33.7%200-13.9%202.8-19.5%2016.3-19.5%2010.9%200%2038.2%200%2038.2%200V82.9c0%200-40.2%200-48.8%200%20-52.5%200-76.1 ...
+  background-image: url('data:image/svg+xml;charset=utf-8,%3Csvg%20xmlns%3D%22http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg%22%20viewBox%3D%220%200%20512%20512%22%3E%3Cpath%20fill%3D%22%23ffffff%22%20d%3D%22M211.9%20197.4h-36.7v59.9h36.7V433.1h70.5V256.5h49.2l5.2-59.1h-54.4c0%200%200-22.1%200-33.7%200-13.9%202.8-19.5%2016.3-19.5%2010.9%200%2038.2%200%2038.2%200V82.9c0%200-40.2%200-48.8%200%20-52.5%200-76. ...
```
Lol, you just ran up against my change. You'll have to rebase, sorry about that.

ithinkihaveacat replied:
All done! (And clever git managed to resolve the conflicts itself, too.)
